### PR TITLE
[BEAM-8338] Support ES 7.x for ElasticsearchIO

### PR DESCRIPTION
**Pull request by regfaker**
R: @dmvk

ElasticsearchIO used operations:
reader
`GET /<index>/<type>/_search`
`GET _search/scroll`
`DELETE _search/scroll`
https://www.elastic.co/guide/en/elasticsearch/reference/7.0/search.html
https://www.elastic.co/guide/en/elasticsearch/reference/7.0/search-request-scroll.html

writer
`POST /<index>/<type>/_bulk`
https://www.elastic.co/guide/en/elasticsearch/reference/7.0/docs-bulk.html

`GET` (for version info)

These are not in https://www.elastic.co/guide/en/elasticsearch/reference/7.0/breaking-changes-7.0.html

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by dmvk** [comment]
cc @aromanenko-dev @echauchot

**Comment by regfaker** [comment]
I'll check that :sdks:java:io:elasticsearch-tests:elasticsearch-tests-7:test problem

**Comment by regfaker** [comment]
Yes, I agree with that duplication. It definitely can be done. I could try it but I don't know all the internals about tests mechanism in gradle and how that integration test with use of create_elk_container.sh works (I can run that docker manually).
So maybe for someone experienced in this it would be peace of cake.
In the meantime I will try.

**Comment by regfaker** [comment]
Ok, there will be couple of additional changes because of some v7 changes ...

File: sdks/java/io/elasticsearch-tests/elasticsearch-tests-7/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/*
+Cannot use @RunWith(JUnit4.class) with ESIntegTestCase
+Cannot have @BeforeClass @AfterClass with ESIntegTestCase
+*/
+
+/** Tests for {@link ElasticsearchIO} version 6. */
```
Comment (aromanenko-dev) [inline sdks/java/io/elasticsearch-tests/elasticsearch-tests-7/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java]: Version 7?

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 2]
```diff
@@ -651,7 +651,7 @@ private BoundedElasticsearchSource(
               new BoundedElasticsearchSource(spec, shardId, null, null, null, backendVersion));
         }
         checkArgument(!sources.isEmpty(), "No shard found");
-      } else if (backendVersion == 5 || backendVersion == 6) {
+      } else if (backendVersion == 5 || backendVersion == 6 || backendVersion == 7) {
```
Comment (aromanenko-dev) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: Just `else {...}` ?

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 3]
```diff
@@ -202,7 +202,7 @@ static void checkForErrors(HttpEntity responseEntity, int backendVersion, boolea
         } else {
           if (backendVersion == 2) {
             errorRootName = "create";
-          } else if (backendVersion == 5 || backendVersion == 6) {
+          } else if (backendVersion == 5 || backendVersion == 6 || backendVersion == 7) {
```
Comment (aromanenko-dev) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: Why not just use simple "if/else" condition in this case? Like:
```
errorRootName = (backendVersion == 2) "create" ? "index";
```

**Comment by aromanenko-dev** [comment]
Also, please take a look on failing tests in Java PreCommit

**Comment by regfaker** [comment]
@echauchot At rest level (I mean requests) there is no breaking changes. I found couple of them at RestClient level itself when I upgraded es client dependency and that is what I'm working on now (almost finished).
But at all I don't like this concept of ifing several versions in one class. Wouldn't it be better to have multiple ElasticsearchIO clients and ElasticsearchIO will use the one it needs (this will needs some shading or if it is some other possibility in gradle (I'm not familiar with it)). I think that it is not sustainable in this state (for example if some big breaking change in RestClient will come and some of older versions (v2) will stop work).

**Comment by echauchot** [comment]
I'm not a big fan of ifs either but the aim was to have a single production code (for maintenance) in a single module (for users) . And as there were only some located differences (mainly split), a simple if could do. 
Also please note 2 things: discrete versions are used instead of > x in the code to avoid running on non-tested versions. Also we used a very low level ES client (rest client) which is (was?) the only one compatible with all the versions of ES, specially to have a single production code base of ESIO.

If you want, you can submit a refactor PR to improve it. I'll be happy of it.

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java (hunk 3 above)
Comment (echauchot) [reply sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: better ton avoid running on non-tested versions (if we use >)

**Comment by echauchot** [comment]
@regfaker I'm deeply sorry but the PR below was submitted 1 day before yours and addresses the same (+ the maxRetryTimeout deprecation). 
https://github.com/apache/beam/pull/10010
I need to close one of the PRs and the only valid criteria is the first submitted PR.
I know it can be frustrating, and I'm sorry.
Thanks anyway for your work and I hope you would not be reluctant to submitting other PRs to Beam in the future, the Beam community would love to see your future work.

**Comment by regfaker** [comment]
It's ok. I'm glad that ES 7 support will make it to release. Maybe next time I can help more.

**Pull request closed without merging**